File: k720_golib/k720.go
package k720

import (
	"errors"
	"fmt"
	"strconv"
	"time"

	"go.bug.st/serial"
)

const (
	STX byte = 0x02
	ETX byte = 0x03
	ENQ byte = 0x05
	ACK byte = 0x06
	NAK byte = 0x15
)

const (
	LOG_NONE  int = 0 // iota
	LOG_TRACE int = 1 // iota
	LOG_DEBUG int = 2 // iota
	LOG_INFO  int = 3 // iota
	LOG_WARN  int = 4 // iota
	LOG_ERROR int = 5 // iota
	LOG_FATAL int = 6 // iota
)

var LOG_LEVEL int = LOG_NONE

type State struct {
	Code        int
	Description string
}

var (
	CARD_AT_SENSOR_1_POSITION  State = registerState(0x0001, "Card at sensor 1 position")
	CARD_AT_SENSOR_2_POSITION  State = registerState(0x0002, "Card at sensor 2 position")
	CARD_AT_SENSOR_3_POSITION  State = registerState(0x0004, "Card at sensor 3 position")
	CARD_EMPTY                 State = registerState(0x0008, "Card empty")
	CARD_PRE_EMPTY             State = registerState(0x0010, "Card pre-empty")
	CARD_JAM                   State = registerState(0x0020, "Card jam")
	CARD_OVERLAP               State = registerState(0x0040, "Card overlap")
	CARD_HOPPER_FULL           State = registerState(0x0080, "Card hopper full")
	ERROR_OF_RECYCLING_CARD    State = registerState(0x0100, "Error of recycling card")
	ERROR_OF_ISSUING_CARD      State = registerState(0x0200, "Error of issuing card")
	COLLECTING_CARD            State = registerState(0x0400, "Collecting card")
	SENDING_CARD               State = registerState(0x0800, "Sending card")
	PREPARING_CARD             State = registerState(0x1000, "Preparing card")
	PREPARE_CARD_FAILURE       State = registerState(0x2000, "Prepare card failure")
	COULD_NOT_IMPLEMNT_COMMAND State = registerState(0x4000, "Could not implement command")
	RECYCLING_BOX_FULL         State = registerState(0x8000, "Recycling box fulls")
)
var states []State = make([]State, 0)

func registerState(code int, description string) State {
	state := State{code, description}
	states = append(states, state)
	return state
}

const (
	KEYA byte = 0x30
	KEYB byte = 0x31
)

func logTrace(args ...interface{}) {
	if LOG_LEVEL == LOG_NONE {
		return
	}
	if LOG_LEVEL <= LOG_TRACE {
		fmt.Println(args...)
	}
}

func logDebug(args ...interface{}) {
	if LOG_LEVEL == LOG_NONE {
		return
	}
	if LOG_LEVEL <= LOG_DEBUG {
		fmt.Println(args...)
	}
}

func debugPacket(prefix string, packet []byte) {
	if LOG_LEVEL == LOG_NONE {
		return
	}
	if LOG_LEVEL <= LOG_DEBUG {
		PrintPacket(prefix, packet)
	}
}

func calculateBcc(packet []byte) byte {
	xorsum := byte(0x00)
	for _, b := range packet {
		xorsum ^= b
	}
	return xorsum
}

func hexEncodeBytes(bytes []byte) string {
	str := ""
	for _, b := range bytes {
		str += fmt.Sprintf("0x%02x ", b)
	}
	return str
}

func commRead(comHandle serial.Port, p []byte) (n int, err error) {
	i := 0
	buf := make([]byte, 1)
	for {
		buf[0] = 0x00
		n, err := comHandle.Read(buf)
		if err != nil {
			return i, err
		}
		if n == 0 {
			return i, nil
		}
		// copy(p[i:], buf[:])
		copy(p[i:], buf)
		i += n
		if i == len(p) {
			return i, nil
		}
	}
}

func createPacket(macAddr byte, data string) ([]byte, error) {
	packet := make([]byte, 0)
	selen := len(data)
	if /*macAddr < 0 ||*/ macAddr > 15 {
		return nil, errors.New("error #1 in createPacket")
	}
	if selen == 0 {
		return nil, errors.New("error #2 in createPacket")
	}

	// start
	packet = append(packet, STX)

	// address
	addrH := int('0') + int(macAddr/10)
	addrL := int('0') + int(macAddr%10)
	packet = append(packet, byte(addrH))
	packet = append(packet, byte(addrL))

	// selen
	selenL := selen & 0xff
	selenH := (selen << 8) & 0xff
	packet = append(packet, byte(selenH))
	packet = append(packet, byte(selenL))
	// data
	packet = append(packet, data...)

	// end
	packet = append(packet, ETX)

	// bcc
	bcc := calculateBcc(packet)
	packet = append(packet, bcc)

	return packet, nil
}

func createEnquiryPacket(macAddr byte) ([]byte, error) {
	packet := make([]byte, 0)
	if /*macAddr < 0 ||*/ macAddr > 15 {
		return nil, errors.New("error #1 in createEnquiryPacket")
	}
	// enquiry
	packet = append(packet, ENQ)

	// address
	addrH := int('0') + int(macAddr/10)
	addrL := int('0') + int(macAddr%10)
	packet = append(packet, byte(addrH))
	packet = append(packet, byte(addrL))

	return packet, nil
}

func readAck(comHandle serial.Port, macAddr byte) error {
	arrAck := make([]byte, 1)
	n, err := commRead(comHandle, arrAck)
	if err != nil {
		return err
	}
	if n != 1 {
		return errors.New("error #1 in readAck")
	}
	ack := arrAck[0]
	if ack != ACK {
		return errors.New("error #2 in readAck")
	}

	arrAddr := make([]byte, 2)
	n, err = commRead(comHandle, arrAddr)
	if err != nil {
		return err
	}
	if n != 2 {
		return errors.New("error #3 in readAck")
	}
	logTrace("Received address:", string(arrAddr))
	if string(arrAddr) != strconv.Itoa(int(macAddr)) {
		return errors.New("error #4 in readAck")
	}
	return nil
}

func sendPacket(comHandle serial.Port, macAddr byte, command string) ([]byte, error) {
	packet, err := createPacket(macAddr, command)
	debugPacket("sent: ", packet)
	if err != nil {
		return nil, err
	}
	comHandle.Write(packet)

	err = readAck(comHandle, macAddr)
	if err != nil {
		return nil, err
	}

	packet, err = createEnquiryPacket(macAddr)
	if err != nil {
		return nil, err
	}
	comHandle.Write(packet)

	packet, err = receivePacket(comHandle, macAddr)
	if err != nil {
		return nil, err
	}
	return packet, nil
}

func receivePacket(comHandle serial.Port, macAddr byte) ([]byte, error) {
	packet := make([]byte, 0)
	// start
	arrStx := make([]byte, 1)
	// n, err := commRead(comHandle, arrStx)
	n, err := comHandle.Read(arrStx)
	logTrace("bSTX: " + hexEncodeBytes(arrStx))
	if err != nil {
		return nil, err
	}
	if n != 1 {
		return nil, errors.New("error #1 in receivePacket")
	}
	stx := arrStx[0]
	if stx != STX {
		return nil, errors.New("error #2 in receivePacket")
	}
	packet = append(packet, stx)

	// address
	arrAddr := make([]byte, 2)
	n, err = commRead(comHandle, arrAddr)
	logTrace("bADDR: " + hexEncodeBytes(arrAddr))
	if err != nil {
		return nil, err
	}
	if n != 2 {
		return nil, errors.New("error #3 in receivePacket")
	}
	if string(arrAddr) != strconv.Itoa(int(macAddr)) {
		return nil, errors.New("error #4 in receivePacket")
	}
	packet = append(packet, arrAddr[0])
	packet = append(packet, arrAddr[1])

	// relen
	arrRelen := make([]byte, 2)
	n, err = commRead(comHandle, arrRelen)
	logTrace("bRELEN: " + hexEncodeBytes(arrRelen))
	if err != nil {
		return nil, err
	}
	if n != 2 {
		return nil, errors.New("error #5 in receivePacket")
	}
	relenH := arrRelen[0]
	relenL := arrRelen[1]
	relen := 0xff*relenH + relenL
	packet = append(packet, relenH)
	packet = append(packet, relenL)
	// data
	arrData := make([]byte, relen)
	n, err = commRead(comHandle, arrData)
	logTrace("bDATA: " + hexEncodeBytes(arrData))
	if err != nil {
		return nil, err
	}
	if n != int(relen) {
		return nil, errors.New("error #6 in receivePacket")
	}
	packet = append(packet, arrData...)

	// end
	arrEtx := make([]byte, 1)
	n, err = commRead(comHandle, arrEtx)
	logTrace("bETX: " + hexEncodeBytes(arrEtx))
	if err != nil {
		return nil, err
	}
	if n != 1 {
		return nil, errors.New("error #7 in receivePacket")
	}
	etx := arrEtx[0]
	if etx != ETX {
		return nil, errors.New("error #8 in receivePacket")
	}
	packet = append(packet, etx)

	// bcc
	bccEtx := make([]byte, 1)
	n, err = commRead(comHandle, bccEtx)
	logTrace("bBCC: " + hexEncodeBytes(bccEtx))
	if err != nil {
		return nil, err
	}
	if n != 1 {
		return nil, errors.New("error #9 in receivePacket")
	}
	bcc := bccEtx[0]
	if bcc != calculateBcc(packet) {
		return nil, errors.New("error #10 in receivePacket")
	}
	packet = append(packet, bcc)

	debugPacket("received: ", packet)
	return packet, nil
}

func getPacketData(packet []byte) ([]byte, error) {
	if len(packet) < 5 {
		return nil, errors.New("error #1 in getPacketData")
	}

	// relen
	relenH := packet[3]
	relenL := packet[4]
	relen := 0xff*relenH + relenL
	// data
	data := packet[5 : 5+relen]
	return data, nil
}

/////////////////////////////////////////////////////////////////////////////

func PrintPacket(prefix string, packet []byte) {
	fmt.Println(prefix + hexEncodeBytes(packet))
}

func CalculateState(query []byte) int {
	state := 0x0000
	for _, b := range query {
		s := b - byte('0')
		state <<= 4
		state |= int(s)
	}
	return state
}

func PrintState(state int) {
	for _, s := range states {
		if state&s.Code > 0 {
			fmt.Println(s.Description)
		}
	}
}

func CommOpen(port string) (serial.Port, error) {
	mode := &serial.Mode{
		BaudRate: 9600,
		DataBits: 8,
		Parity:   serial.NoParity,
		StopBits: serial.OneStopBit,
	}
	comHandle, err := serial.Open(port, mode)
	comHandle.SetReadTimeout(1 * time.Second)
	return comHandle, err
}

func CommOpenWitBaud(port string, baudrate int) (serial.Port, error) {
	mode := &serial.Mode{
		BaudRate: baudrate,
		DataBits: 8,
		Parity:   serial.NoParity,
		StopBits: serial.OneStopBit,
	}
	comHandle, err := serial.Open(port, mode)
	comHandle.SetReadTimeout(1 * time.Second)
	return comHandle, err
}

func CommClose(comHandle serial.Port) error {
	return comHandle.Close()
}

func GetSysVersion(comHandle serial.Port, macAddr byte) (string, error) {
	packet, err := sendPacket(comHandle, macAddr, "GV")
	if err != nil {
		return "", err
	}
	data, err := getPacketData(packet)
	if err != nil {
		return "", err
	}
	return string(data), nil
}

func Query(comHandle serial.Port, macAddr byte) ([]byte, error) {
	packet, err := sendPacket(comHandle, macAddr, "RF")
	if err != nil {
		return nil, err
	}
	query, err := getPacketData(packet)
	if err != nil {
		return nil, err
	}
	if len(query) < 3 {
		return nil, errors.New("error #1 in Query")
	}
	return query[2:], nil
}

func SensorQuery(comHandle serial.Port, macAddr byte) ([]byte, error) {
	packet, err := sendPacket(comHandle, macAddr, "AP")
	if err != nil {
		return nil, err
	}
	query, err := getPacketData(packet)
	if err != nil {
		return nil, err
	}
	if len(query) < 3 {
		return nil, errors.New("error #1 in SensorQuery")
	}
	return query[2:], nil
}

func SendCmd(comHandle serial.Port, macAddr byte, command string) ([]byte, error) {
	packet, err := sendPacket(comHandle, macAddr, command)
	if err != nil {
		return nil, err
	}
	return getPacketData(packet)
}

func S50DetectCard(comHandle serial.Port, macAddr byte) ([]byte, error) {
	packet, err := sendPacket(comHandle, macAddr, "\x3b\x30")
	if err != nil {
		return nil, err
	}
	return getPacketData(packet)
}

func S50GetCardId(comHandle serial.Port, macAddr byte) ([]byte, error) {
	packet, err := sendPacket(comHandle, macAddr, "\x3b\x31")
	if err != nil {
		return nil, err
	}
	return getPacketData(packet)
}

func S50LoadSecKey(comHandle serial.Port, macAddr byte, sectorAddr byte, keyType byte, key []byte) ([]byte, error) {
	command := "\x3b\x32" +
		string(sectorAddr) +
		string(keyType) +
		string(key)
	packet, err := sendPacket(comHandle, macAddr, command)
	if err != nil {
		return nil, err
	}
	return getPacketData(packet)
}

func S70DetectCard(comHandle serial.Port, macAddr byte) ([]byte, error) {
	packet, err := sendPacket(comHandle, macAddr, "\x3c\x30")
	if err != nil {
		return nil, err
	}
	return getPacketData(packet)
}

func S70GetCardId(comHandle serial.Port, macAddr byte) ([]byte, error) {
	packet, err := sendPacket(comHandle, macAddr, "\x3c\x31")
	if err != nil {
		return nil, err
	}
	return getPacketData(packet)
}

func ULDetectCard(comHandle serial.Port, macAddr byte) ([]byte, error) {
	packet, err := sendPacket(comHandle, macAddr, "\x3d\x30")
	if err != nil {
		return nil, err
	}
	return getPacketData(packet)
}

func ULGetCardId(comHandle serial.Port, macAddr byte) ([]byte, error) {
	packet, err := sendPacket(comHandle, macAddr, "\x3d\x31")
	if err != nil {
		return nil, err
	}
	return getPacketData(packet)
}

File: k720_pylib/k720.py
import serial

STX = 0x02
ETX = 0x03
ENQ = 0x05
ACK = 0x06
NAK = 0x15

LOG_NONE, LOG_TRACE, LOG_DEBUG, LOG_INFO, LOG_WARN, LOG_ERROR, LOG_FATAL = \
    (0, 1, 2, 3, 4, 5, 6)
LOG_LEVEL = LOG_NONE

CARD_AT_SENSOR_1_POSITION = 0x0001
CARD_AT_SENSOR_2_POSITION = 0x0002
CARD_AT_SENSOR_3_POSITION = 0x0004

KEYA = 0x30
KEYB = 0x31

def _log_trace(*objects):
    if LOG_LEVEL == LOG_NONE:
        return
    if LOG_LEVEL <= LOG_TRACE:
        print(*objects)

def _log_debug(*objects):
    if LOG_LEVEL == LOG_NONE:
        return
    if LOG_LEVEL <= LOG_DEBUG:
        print(*objects)

def _debug_packet(prefix, packet):
    if LOG_LEVEL == LOG_NONE:
        return
    if LOG_LEVEL <= LOG_DEBUG:
        print_packet(prefix, packet)

def _calculate_bcc(packet):
    # type: (list) -> int
    xorsum = 0x00
    for b in packet:
        xorsum ^= b
    return xorsum

def _create_packet(mac_addr, data):
    # type: (int, str) -> list | None
    packet = [] 
    selen = len(data)
    if mac_addr < 0 or mac_addr > 15:
        return None
    if selen == 0:
        return None

    # start
    packet.append(STX)

    # address
    addr_h = ord('0') + mac_addr // 10
    addr_l = ord('0') + mac_addr % 10
    packet.append(addr_h)
    packet.append(addr_l)

    # selen
    selen_l = selen & 0xff
    selen_h = (selen << 8) & 0xff
    packet.append(selen_h)
    packet.append(selen_l)
    # data
    for d in data:
        packet.append(ord(d))

    # end
    packet.append(ETX)

    # bcc
    bcc = _calculate_bcc(packet)
    packet.append(bcc)

    return packet

def _create_enquiry_packet(mac_addr):
    # type: (int) -> list | None 
    packet = []
    if mac_addr < 0 or mac_addr > 15:
        return None
    # enquiry
    packet.append(ENQ)

    # address
    addr_h = ord('0') + (mac_addr // 10)
    addr_l = ord('0') + (mac_addr % 10)
    packet.append(addr_h)
    packet.append(addr_l)

    return packet

def _read_ack(com_handle, mac_addr):
    # type: (object, int) -> bool
    b_ack = com_handle.read(size=1)
    if len(b_ack) != 1:
        return False
    ack = b_ack[0]    
    if ack != ACK:
        return False

    b_addr = com_handle.read(size=2)
    _log_trace("Received address:", b_addr.decode())
    if b_addr.decode() != str(mac_addr):
        return False
    return True

def _send_packet(com_handle, mac_addr, command):
    # type: (object, int, str) -> list | None
    packet = _create_packet(mac_addr, command)
    if packet == None:
        return None
    _debug_packet('sent: ', packet)
    com_handle.write(packet)

    if not _read_ack(com_handle, mac_addr):
        return None

    packet = _create_enquiry_packet(mac_addr)
    if packet == None:
        return None
    com_handle.write(packet)

    packet = _receive_packet(com_handle, mac_addr)
    if packet == None:
        return None
    return packet

def _receive_packet(com_handle, mac_addr):
    # type: (object, int) -> list | None
    packet = []
    # start
    b_stx = com_handle.read(size=1)
    _log_trace("bSTX:" + str(bytes(b_stx)))
    if len(b_stx) != 1:
        return None
    stx = b_stx[0]
    if stx != STX:
        return None
    packet.append(stx)

    # address
    b_addr = com_handle.read(size=2)
    _log_trace("bADDR:" + str(bytes(b_addr)))
    if len(b_addr) != 2:
        return None
    if b_addr.decode() != str(mac_addr):
        return None
    packet.append(b_addr[0])
    packet.append(b_addr[1])

    # relen
    b_relen = com_handle.read(size=2)
    _log_trace("bRELEN:" + str(bytes(b_relen)))
    if len(b_relen) != 2:
        return None
    relen_h = b_relen[0]
    relen_l = b_relen[1]
    relen = 0xff * relen_h + relen_l
    packet.append(relen_h)
    packet.append(relen_l)
    # data
    b_data = com_handle.read(size=relen)
    _log_trace("bDATA:" + str(bytes(b_data)))
    if len(b_data) != relen:
        return None
    for d in b_data:
        packet.append(d)

    # end
    b_etx = com_handle.read(size=1)
    _log_trace("bETX:" + str(bytes(b_etx)))
    if len(b_etx) != 1:
        return None
    etx = b_etx[0]
    if etx != ETX:
        return None
    packet.append(etx)
    
    # bcc
    b_bcc = com_handle.read(size=1)
    _log_trace("bBCC:" + str(bytes(b_bcc)))
    if len(b_bcc) != 1:
        return None
    bcc = b_bcc[0]
    if bcc != _calculate_bcc(packet):
        return None
    packet.append(bcc)

    _debug_packet('received: ', packet)
    return packet

def _get_packet_data(packet):
    # type: (list) -> list | None
    data = []
    if len(packet) < 5:
        return None

    # relen
    relen_h = packet[3]
    relen_l = packet[4]
    relen = 0xff * relen_h + relen_l
    # data
    data = packet[5:5+relen]
    return data

#############################################################################

def print_packet(prefix, packet):
    # type: (str, list) -> None
    hex_sequence = ' '.join(['0x' + format(b, '02x') for b in packet])
    print(prefix + hex_sequence)

def calculate_state(query):
    # type: (list) -> int
    state = 0x0000
    for b in query:
        s = b - ord('0')
        state <<= 4
        state |= s
    # print(hex(state))
    return state

def comm_open(port):
    # type: (str) -> object
    return serial.Serial(
        port=port,
        baudrate=9600,
        bytesize=serial.EIGHTBITS,
        parity=serial.PARITY_NONE,
        stopbits=serial.STOPBITS_ONE,
        timeout=1)

def comm_open_with_baud(port, baudrate):
    # type: (str, int) -> object
    return serial.Serial(
        port=port,
        baudrate=baudrate,
        bytesize=serial.EIGHTBITS,
        parity=serial.PARITY_NONE,
        stopbits=serial.STOPBITS_ONE,
        timeout=1)

def comm_close(com_handle):
    # type: (object) -> None
    com_handle.close()

# Get Version (GV)
def get_sys_version(com_handle, mac_addr):
    # type: (object, int) -> str | None
    packet = _send_packet(com_handle, mac_addr, 'GV')
    if packet == None:
        return None
    data = _get_packet_data(packet)
    if data == None:
        return None
    return bytes(data).decode()

# Check Sensor (RF)
def query(com_handle, mac_addr):
    # type: (object, int) -> list | None
    packet = _send_packet(com_handle, mac_addr, 'RF')
    if packet == None:
        return None
    query = _get_packet_data(packet)
    if len(query) < 3:
        return None
    return query[2:]

# Advanced Check (AP)
def sensor_query(com_handle, mac_addr):
    # type: (object, int) -> list | None
    packet = _send_packet(com_handle, mac_addr, 'AP')
    if packet == None:
        return None
    sensor_query = _get_packet_data(packet)
    if len(sensor_query) < 3:
        return None
    return sensor_query[2:]

def send_cmd(com_handle, mac_addr, command):
    # type: (object, int, str) -> list | None
    packet = _send_packet(com_handle, mac_addr, command)
    if packet == None:
        return None
    return _get_packet_data(packet)

def s50_detect_card(com_handle, mac_addr):
    # type: (object, int) -> list | None
    packet = _send_packet(com_handle, mac_addr, '\x3b\x30')
    if packet == None:
        return None
    return _get_packet_data(packet)

def s50_get_card_id(com_handle, mac_addr):
    # type: (object, int) -> list | None
    packet = _send_packet(com_handle, mac_addr, '\x3b\x31')
    if packet == None:
        return None
    return _get_packet_data(packet)

def s50_load_sec_key(com_handle, mac_addr, sector_addr, key_type, key):
    # type: (object, int, int, int, list) -> list | None
    command = '\x3b\x32' + \
            chr(sector_addr) + \
            chr(key_type) + \
            ''.join([chr(b) for b in key])
    packet = _send_packet(com_handle, mac_addr, command)
    if packet == None:
        return None
    return _get_packet_data(packet)

def s70_detect_card(com_handle, mac_addr):
    # type: (object, int) -> list | None
    packet = _send_packet(com_handle, mac_addr, '\x3c\x30')
    if packet == None:
        return None
    return _get_packet_data(packet)

def s70_get_card_id(com_handle, mac_addr):
    # type: (object, int) -> list | None
    packet = _send_packet(com_handle, mac_addr, '\x3c\x31')
    if packet == None:
        return None
    return _get_packet_data(packet)

def ul_detect_card(com_handle, mac_addr):
    # type: (object, int) -> list | None
    packet = _send_packet(com_handle, mac_addr, '\x3d\x30')
    if packet == None:
        return None
    return _get_packet_data(packet)

def ul_get_card_id(com_handle, mac_addr):
    # type: (object, int) -> list | None
    packet = _send_packet(com_handle, mac_addr, '\x3d\x31')
    if packet == None:
        return None
    return _get_packet_data(packet)
